### PR TITLE
[WIP] Kvcache fix

### PR DESCRIPTION
**Pull request by hexisyztem**
Remove all original concatenation handling of kv cache to avoid unnecessary processing of long length kv. Preallocate the required space for kv cache so that it is not necessary to reallocate kv cache each time a new step result is filled in.

### CODE REVIEW

**Comment by hexisyztem** [comment]
change target branch.

**Comment by hexisyztem** [comment]
Related issue: https://github.com/FunAudioLLM/CosyVoice/issues/184

File: cosyvoice/llm/llm.py [hunk 1]
```diff
@@ -192,10 +192,25 @@ def inference(
         # 5. step by step decode
         out_tokens = []
         offset = 0
-        att_cache, cnn_cache = torch.zeros((0, 0, 0, 0), device=lm_input.device), torch.zeros((0, 0, 0, 0), device=lm_input.device)
+        # Note: key cache shape is (batch, nhead, max_seq, head_dim). Preallocate the entire tensor in advance,
+        #  and for step-k, fill the key into key_cache[:, :, step_k, :]; the value cache follows the same logic.
+        key_caches = [torch.zeros((llm_input.size(0), self.llm.attention_heads, max_len, self.llm.head_dim), device=lm_input.device) for _ in range(len(self.llm.encoders))]
```
Comment (aluminumbox) [inline cosyvoice/llm/llm.py:197]: 这个key_cache、value_cache不要分两个，之前att_cache的shape是14, 16，n, 128，那么还是叫att_cache，只不过是一个list，然后list大小是14, 然后list的每个元素是一个大小为2的list，存的分别是每一层的key/value， 大小是1, 16, max_len, 64，这样的话，接口就不用改变了

File: cosyvoice/transformer/attention.py [hunk 2]
```diff
@@ -73,9 +75,6 @@ def forward_qkv(
         q = self.linear_q(query).view(n_batch, -1, self.h, self.d_k)
         k = self.linear_k(key).view(n_batch, -1, self.h, self.d_k)
         v = self.linear_v(value).view(n_batch, -1, self.h, self.d_k)
-        q = q.transpose(1, 2)  # (batch, head, time1, d_k)
-        k = k.transpose(1, 2)  # (batch, head, time2, d_k)
-        v = v.transpose(1, 2)  # (batch, head, time2, d_k)
```
Comment (aluminumbox) [inline cosyvoice/transformer/attention.py:78]: 这个transpose的位置移动有讲究吗？

File: cosyvoice/transformer/attention.py [hunk 3]
```diff
@@ -183,18 +187,38 @@ def forward(
...
-                                                 dim=-1)
-            k = torch.cat([key_cache, k], dim=2)
-            v = torch.cat([value_cache, v], dim=2)
-        # NOTE(xcsong): We do cache slicing in encoder.forward_chunk, since it's
-        #   non-trivial to calculate `next_cache_start` here.
-        new_cache = torch.cat((k, v), dim=-1)
 
-        scores = torch.matmul(q, k.transpose(-2, -1)) / math.sqrt(self.d_k)
-        return self.forward_attention(v, scores, mask), new_cache
+        if key_cache is not None and value_cache is not None:
```
Comment (aluminumbox) [inline cosyvoice/transformer/attention.py:191]: 这里也是一样，如果之前的接口保持一致，是一个len 14的list，list里是一个len 2的list分别存储key value，那么这里的判断应该也就不需要太多变化了

File: cosyvoice/transformer/attention.py [hunk 4]
```diff
@@ -288,15 +316,21 @@ def forward(
...
+            key_len, value_len = cache_offset + k.size(2), cache_offset + v.size(2)
+            # NOTE: This will cause the torch.tensor to be non-contiguous. However, since it 
+            # participates in matmul calculations and the subsequent transpose operations do 
+            # not change the memory values, only the stride, there is no need to create a new 
+            # tensor using contiguous here, as the kv cache requires a large amount of space.
+            # TODO: It is necessary to verify whether executing contiguous can bring performance benefits.
+            key = key_cache[:, :, :key_len, :]
+            value = value_cache[:, :, :value_len, :]
+        else:
+            key = k
```
Comment (aluminumbox) [inline cosyvoice/transformer/attention.py:332]: 下面还是有k这个变量吧，逻辑上也可以和之前保持一致，不用执行else里的内容

File: cosyvoice/transformer/encoder.py [hunk 5]
```diff
@@ -231,45 +235,46 @@ def forward_chunk(
...
+
         r_cnn_cache = []
         for i, layer in enumerate(self.encoders):
             # NOTE(xcsong): Before layer.forward
-            #   shape(att_cache[i:i + 1]) is (1, head, cache_t1, d_k * 2),
+            #   shape(key_caches[i:i + 1]) is (1, head, cache_offset, d_k),
+            #   shape(value_caches[i:i + 1]) is (1, head, cache_offset, d_k),
             #   shape(cnn_cache[i])       is (b=1, hidden-dim, cache_t2)
-            xs, _, new_att_cache, new_cnn_cache = layer(
+            xs, _, key_caches[i], value_caches[i], new_cnn_cache = layer(
```
Comment (aluminumbox) [inline cosyvoice/transformer/encoder.py:256]: 这里也是一样，r_att_cache是一个大小14的list，这个循环逻辑应该可以保持不变

File: cosyvoice/llm/llm.py [hunk 6]
```diff
@@ -192,10 +192,25 @@ def inference(
...
+        value_caches = [torch.zeros((llm_input.size(0), self.llm.attention_heads, max_len, self.llm.head_dim), device=lm_input.device) for _ in range(len(self.llm.encoders))]
```
Comment (aluminumbox) [inline cosyvoice/llm/llm.py:198]: 这个max_len应该是5000，是pos_embed的最大长度

File: cosyvoice/transformer/attention.py [hunk 7]
```diff
@@ -311,16 +345,35 @@ def forward(
...
 
-        scores = (matrix_ac + matrix_bd) / math.sqrt(
-            self.d_k)  # (batch, head, time1, time2)
+        if not self.use_sdpa:
+            scores = (matrix_ac + matrix_bd) / math.sqrt(
+                self.d_k)  # (batch, head, time1, time2)
 
-        return self.forward_attention(v, scores, mask), new_cache
+            return self.forward_attention(value, scores, mask), key_cache, value_cache
+        else:
```
Comment (shanhaidexiamo) [inline cosyvoice/transformer/attention.py:363]: 请教一下else这里操作的依据是什么，为什么没用到前面计算的matrix_ac呢

Comment (shanhaidexiamo) [inline cosyvoice/transformer/attention.py:363]: 前面代码matrix_ac 计算了一次大矩阵乘，在else这儿又要通过scaled_dot_product_attention计算一次乘，再算上matrix_bd，这块要计算三个大矩阵乘法了。

Comment (hexisyztem) [reply cosyvoice/transformer/attention.py:363]: 是的，原本的 attention 是需要计算 score = q @ k 和 output = score @ v，但是 RelPositionMultiHeadedAttention 这块的计算逻辑是需要计算 q @ k 作为bias加入到 mask 中。从计算机制上来说没法避免吧。

Comment (hexisyztem) [reply cosyvoice/transformer/attention.py:362]: 这里的if else是对同一个计算逻辑的两种不同实现方式。sdpa 是使用的 fuse kernel，性能更高效一些。理论上应该使用 flash attn的，但是因为 flash attn 不支持 mask，所以这里暂时没有用，后面再看看。

Comment (shanhaidexiamo) [reply cosyvoice/transformer/attention.py:362]: 但是似乎matrix_ac的计算可以在else的情况下避免，只需要计算出来matrix_ac的shape即可，这样就只需要计算matrix_bd？

File: cosyvoice/transformer/attention.py (hunk 2 above)
Comment (hexisyztem) [reply cosyvoice/transformer/attention.py:78]: 因为 RelPositionMultiHeadedAttention 中 q 又转置回原本的shape了

File: cosyvoice/llm/llm.py (hunk 1 above)
Comment (hexisyztem) [reply cosyvoice/llm/llm.py]: Done

File: cosyvoice/transformer/attention.py (hunk 3 above)
Comment (hexisyztem) [reply cosyvoice/transformer/attention.py]: Done

File: cosyvoice/transformer/attention.py (hunk 4 above)
Comment (hexisyztem) [reply cosyvoice/transformer/attention.py]: Done

File: cosyvoice/transformer/attention.py (hunk 7 above)
Comment (hexisyztem) [reply cosyvoice/transformer/attention.py:362]: 感谢提醒，这里是我代码写错了。scores 的计算是无论if还是else都需要的

Comment (hexisyztem) [reply cosyvoice/transformer/attention.py:362]: 感谢提醒，确实是的

Comment (hexisyztem) [reply cosyvoice/transformer/attention.py:362]: > 前面代码matrix_ac 计算了一次大矩阵乘，在else这儿又要通过scaled_dot_product_attention计算一次乘，再算上matrix_bd，这块要计算三个大矩阵乘法了。

scaled_dot_product_attention 计算的是 Softmax(q @ k * scale + mask) @ v，这里面本身就包含了两次矩阵乘法，所以其实是做了四次，计算逻辑上需要三次，确实多了一次。

Comment (shanhaidexiamo) [reply cosyvoice/transformer/attention.py:362]: 请教一下您的torch版本，我是2.0， scaled_dot_product_attention没有scale参数，运行之后的logp全是nan

Comment (hexisyztem) [reply cosyvoice/transformer/attention.py:363]: 这个应该得 2.2.0+ 的torch 版本，它是把 Softmax(q @ k * scale + mask) @ v 这部分计算 fuse 成了一个 kernel
下面这里是我对 scaled_dot_product_attention 运算进行压测做 profile
image

Comment (shanhaidexiamo) [reply cosyvoice/transformer/attention.py:363]: 感谢，我这套流程走下来，是mask 本身全NAN，还没走到SDPA，我再check一下。另外，传参里的dropout_p=self.dropout_rate,  后者应该是不存在的，需要在init里补一个~

Comment (hexisyztem) [reply cosyvoice/transformer/attention.py:363]: 看了一下是我写的代码问题，因为我测试不是对这个 commit 进行的，我后续补一下测试数据

Comment (Pydataman) [reply cosyvoice/transformer/attention.py:367]: @hexisyztem 我这里用了这个代码 生成语音不对
